src/vs/server/node/auth-http.ts

```diff
@@ -0,0 +1,70 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Gitpod. All rights reserved.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as path from 'path';
+import * as http from 'http';
+import { ILogService } from 'vs/platform/log/common/log';
+import { parse } from 'querystring';
+import { args } from 'vs/server/node/args';
+import { authenticated, generateAndSetPassword, handlePasswordValidation } from 'vs/server/node/auth';
+import { APP_ROOT, serveError, serveFile } from 'vs/server/node/server.main';
+
+const LOGIN = path.join(APP_ROOT, 'out', 'vs', 'server', 'browser', 'workbench', 'login.html');
+
+export async function handleServerRequest(pathname: string | null, req: http.IncomingMessage, res: http.ServerResponse, logService: ILogService): Promise<boolean> {
+	if (args.password === undefined) {
+		await generateAndSetPassword(logService);
+	}
+	const auth = await authenticated(args, req);
+	if (!auth) {
+		if (pathname === '/') {
+			serveFile(logService, req, res, LOGIN);
+			return true;
+		}
+		if (pathname === '/login') {
+			const password = (await collectRequestData(req)).password;
+			const { isPasswordValid, hashedPassword } = await handlePasswordValidation({
+				passwordFromRequestBody: password,
+				passwordFromArgs: args.password,
+			});
+
+			if (isPasswordValid) {
+				res.writeHead(302, {
```

Comment:
can we do without a redirect?

I would imagine that we add a new callback like `verifyClient` which is used for all incoming requests and returns boolean value whether we need to continue or not. We don't need to reuse `handleRequest` for it. This function will be called in ws upgrade callback as well as in http listener.

Reply:
I added a new verification handler and did it without the redirect. The only caveat is that when users login and try to reload the app, the browser will ask to resend form information.